pinot-core/src/main/java/org/apache/pinot/core/data/table/ConcurrentSortedIndexedTable.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.table;
+
+import java.util.Comparator;
+import java.util.TreeMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.query.request.context.QueryContext;
+
+
+/**
+ * {@link Table} thread-safe implementation for orderby group key case, sorted on group keys and only `trimSize` groups
+ * are kept
+ */
+@NotThreadSafe
+public class ConcurrentSortedIndexedTable extends IndexedTable {
+  Comparator<Key> _keyComparator;
+  ReadWriteLock _lock = new ReentrantReadWriteLock();
+
+  public ConcurrentSortedIndexedTable(DataSchema dataSchema, boolean hasFinalInput,
+      QueryContext queryContext, int resultSize, ExecutorService executorService, Comparator<Key> keyComparator) {
+    super(dataSchema, hasFinalInput, queryContext, resultSize, resultSize, Integer.MAX_VALUE,
+        new TreeMap<>(keyComparator),
+        executorService);
+    _keyComparator = keyComparator;
+  }
+
+  @Override
+  public boolean upsert(Key key, Record record) {
```

Comment:
I'd be good to make sure locks are released with `try {} finally {}` blocks, otherwise an OOME might  cause one thread to keep write lock indefinitely and lock all other threads participating in combine stage (on `_lock.writeLock().lock()`).